# Spam filter For SMS Messages Using Naive Bayes Algorithm
In this project, we will develop a spam filter for SMS messages using the multinomial Naive Bayes algorithm. Our goal is to build a program that can accurately classify new messages, therefore we expect that more than 80% of new messages will be appropriately classified as either spam or ham (non-spam).

To train the algorithm, we will use a dataset of 5,572 SMS messages that humans have already classified. The dataset was created by Tiago A. Almeida and José María Gómez Hidalgo and can be downloaded from [The UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection). More information about the data collection procedure can be found on [this website](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition), along with papers by Tiago A. Almeida and José María Gómez Hidalgo.

# Data Exploration
We'll start by reading in the dataset.

In [1]:
import pandas as pd
sms_messages = pd.read_csv('SMSSpamCollection',sep = '\t', header = None, names =['label', 'SMS'])
sms_messages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   label   5572 non-null   object
 1   SMS     5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [2]:
sms_messages.head() # display first five rows

label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [3]:
sms_messages['label'].value_counts(normalize = True) * 100

ham     86.593683
spam    13.406317
Name: label, dtype: float64

We find that approximately **87%** of the messages are ham(non-spam), with the remaining **13%** being spam. This sample appears representative, as most messages that people get are ham.

# Training and Test Set
We're now going to split our dataset into a training and a test set, where the training set accounts for 80% of the data, and the test set for the remaining 20%.

In [4]:
# Randomize the dataset
data_randomized = sms_messages.sample(frac=1, random_state=1)

# Calculate index for split
training_test_index = round(len(data_randomized) * 0.8)

# Training/Test split
training_set = data_randomized[:training_test_index].reset_index(drop=True)
test_set = data_randomized[training_test_index:].reset_index(drop=True)

# Number of rows and columns in the both datasets
print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


We'll now analyze the percentage of spam and ham messages in the training and test sets. We expect the percentages to be close to what we have in the full dataset, where about 87% of the messages are ham, and the remaining 13% are spam.

In [5]:
training_set['label'].value_counts(normalize= True) * 100

ham     86.54105
spam    13.45895
Name: label, dtype: float64

In [6]:
test_set['label'].value_counts(normalize= True) * 100

ham     86.804309
spam    13.195691
Name: label, dtype: float64

The results appear to be good! We'll now proceed to cleansing the dataset.

# Data Cleaning
To calculate all the probabilities required by the algorithm, we'll first need to perform a bit of data cleaning to bring the data in a format that will allow us to extract easily all the information we need.

Essentially, we want to bring data to this format:

![img](https://dq-content.s3.amazonaws.com/433/cpgp_dataset_3.png)


## Letter Case and Punctuation

We'll begin with removing all the punctuation and bringing every letter to lower case.

In [7]:
# Before cleaning
training_set.head()

label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [8]:
# After Cleaning
# replace words that are neither alphabet nor numbers e.g punctuation marks with empty space
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ') 
training_set['SMS'] = training_set['SMS'].str.lower() # change letters to lower case
training_set.head()

C:\Users\USER\AppData\Local\Temp\ipykernel_11648\3566198951.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')


label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

# Creating Vocabulary
Let us now create the vocabulary, which is a list of all the unique words in our training set.

In [9]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))
 
print(len(vocabulary))

7783


 There are **7,783** unique words in all the messages of our training set.

# The Final Training Set
We will now use the vocabulary we have recently created to do the desired data transformation.

In [10]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [11]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

workand  charity  intrepid  8th  arsenal  dates  too  club4mobiles  ga  \
0        0        0         0    0        0      0    0             0   0   
1        0        0         0    0        0      0    0             0   0   
2        0        0         0    0        0      0    0             0   0   
3        0        0         0    0        0      0    0             0   0   
4        0        0         0    0        0      0    0             0   0   

   dungerees  ...  shu  lanre  lead  thnk  noncomittal  parco  fluids  \
0          0  ...    0      0     0     0            0      0       0   
1          0  ...    0      0     0     0            0      0       0   
2          0  ...    0      0     0     0            0      0       0   
3          0  ...    0      0     0     0            0      0       0   
4          0  ...    0      0     0     0            0      0       0   

   impressed  sac  txting  
0          0    0       0  
1          0    0       0  
2          0    0       0  
3          0    0       0  
4          0    0       0  

[5 rows x 7783 columns]

In [12]:
training_set_clean = pd.concat([training_set, word_counts], axis=1) # add the word_counts dataset to training set dataset 
training_set_clean.head()

label                                                SMS  workand  charity  \
0   ham                  [yep, by, the, pretty, sculpture]        0        0   
1   ham  [yes, princess, are, you, going, to, make, me,...        0        0   
2   ham                    [welp, apparently, he, retired]        0        0   
3   ham                                           [havent]        0        0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...        0        0   

   intrepid  8th  arsenal  dates  too  club4mobiles  ...  shu  lanre  lead  \
0         0    0        0      0    0             0  ...    0      0     0   
1         0    0        0      0    0             0  ...    0      0     0   
2         0    0        0      0    0             0  ...    0      0     0   
3         0    0        0      0    0             0  ...    0      0     0   
4         0    0        0      0    0             0  ...    0      0     0   

   thnk  noncomittal  parco  fluids  impressed  sac  txting  
0     0            0      0       0          0    0       0  
1     0            0      0       0          0    0       0  
2     0            0      0       0          0    0       0  
3     0            0      0       0          0    0       0  
4     0            0      0       0          0    0       0  

[5 rows x 7785 columns]

# Calculating Constants First
We've finished cleaning the training set and can now begin creating the spam filter. To classify new messages, the Naive Bayes algorithm will need to answer the following two probability questions:

\begin{equation}
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam)
\end{equation}

\begin{equation}
P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot \prod_{i=1}^{n}P(w_i|Ham)
\end{equation}


Also, to calculate P(w<sub>i</sub>|Spam) and P(w<sub>i</sub>|Ham) inside the formulas above, we'll need to use these equations:

\begin{equation}
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
\end{equation}

\begin{equation}
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
\end{equation}


Some of the terms in the four equations above will have the same value for every new message. We can calculate the value of these terms once and avoid doing the computations again when a new messages comes in. Below, we'll use our training set to calculate:

- P(Spam) and P(Ham)
- N<sub>Spam</sub>, N<sub>Ham</sub>, N<sub>Vocabulary</sub>

We'll also use Laplace smoothing and set $\alpha = 1$.



In [13]:
# Isolating spam and ham messages first
spam_messages = training_set_clean[training_set_clean['label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean)
p_ham = len(ham_messages) / len(training_set_clean)

# N_Spam
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N_Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

# Calculating Parameters

Now that we have the constant terms calculated above, we can move on with calculating the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

\begin{equation}
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
\end{equation}

\begin{equation}
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
\end{equation}

In [14]:
# Initiate parameters
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()   # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   # ham_messages already defined in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham

# Classifying A New Message
Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter can be understood as a function that:

+ Takes in as input a new message (w1, w2, ..., wn).
+ Calculates P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn).
+ Compares the values of P(Spam|w1, w2, ..., wn) and P(Ham|w1, w2, ..., wn), and:
    + If P(Ham|w1, w2, ..., wn) > P(Spam|w1, w2, ..., wn), then the message is classified as ham.
    + If P(Ham|w1, w2, ..., wn) < P(Spam|w1, w2, ..., wn), then the message is classified as spam.
    + If P(Ham|w1, w2, ..., wn) = P(Spam|w1, w2, ..., wn), then the algorithm may request human help.

In [15]:
import re
def classify(message):
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [16]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [17]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


#  Measuring the Spam Filter's Accuracy
The two results above look promising, but let's see how well the filter does on our test set, which has 1,114 messages.

We'll start by writing a function that returns classification labels instead of printing them.

In [18]:
def classify_test_set(message):    
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

Now that we have a function that returns labels instead of printing them, we can use it to create a new column in our test set.

In [19]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

Now, we'll write a function to measure the accuracy of our spam filter to find out how well our spam filter does.

In [20]:
correct = 0
total = test_set.shape[0]
    
for row in test_set.iterrows():
    row = row[1]
    if row['label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', (correct/total) * 100)

Correct: 1100
Incorrect: 14
Accuracy: 98.74326750448833


The accuracy is close to 98.74%, which is really good. Our spam filter looked at 1,114 messages that it hasn't seen in training, and classified 1,100 correctly.

# Conclusion 
In this project, we managed to build a spam filter for SMS messages using the **multinomial Naive Bayes algorithm**. The filter had an accuracy of **98.74%** on the test set we used, which is a pretty good result. Our initial goal was an accuracy of **over 80%**, and we managed to do way better than that.

Next steps include:

+ Analyze the 14 messages that were classified incorrectly and try to figure out why the algorithm classified them incorrectly
+ Make the filtering process more complex by making the algorithm sensitive to letter case